Firestore/core/src/firebase/firestore/util/status_apple.mm

```diff
@@ -18,30 +18,99 @@
 
 #if defined(__APPLE__)
 
+#include "Firestore/core/src/firebase/firestore/util/error_apple.h"
 #include "Firestore/core/src/firebase/firestore/util/string_format.h"
+#include "absl/memory/memory.h"
 
 namespace firebase {
 namespace firestore {
 namespace util {
 
+class UnderlyingNSError : public PlatformError {
+ public:
+  explicit UnderlyingNSError(NSError* error) : error_(error) {
+  }
+
+  ~UnderlyingNSError() {
+  }
+
+  static std::unique_ptr<UnderlyingNSError> Create(NSError* error) {
+    return absl::make_unique<UnderlyingNSError>(error);
+  }
+
+  static NSError* Recover(const std::unique_ptr<PlatformError>& wrapped) {
+    if (wrapped == nullptr) {
+      return nil;
+    }
+
+    return static_cast<UnderlyingNSError*>(wrapped.get())->error();
```

Comment:
You haven't verified that wrapped is actually of type UnderlyingNSError. You could:
a) Use RTTI. (Except we can't)
b) Add a type() method to PlatformError.
c) Change the wrapped parameter to be a unique_ptr<UnderlyingNSError>. Which mostly works if you refactor Status::CausedBy as suggested, though it leaves Status::ToNSError with the same problem.
d) Not worry about it. *Clearly*, we're dealing with NSErrors here. (As other platforms/error types are added, I believe this will continue to be the case too; I don't think we'll ever have a state with more than one subtype of PlatformError active at a time.)

No action required.